uber/models.py

```diff
@@ -494,93 +494,138 @@ def lookup_attendee(self, full_name, email, zip_code):
-            next = self.next_badge_num(badge_type, old_badge_num)
-            if out_of_range:
-                return out_of_range
-            elif not badge_num and next > c.BADGE_RANGES[badge_type][1]:
-                return 'There are no more badges available for that type'
-            elif badge_type in c.PREASSIGNED_BADGE_TYPES and not c.SHIFT_CUSTOM_BADGES:
-                return 'Custom badges have already been ordered'
-
-            if not c.SHIFT_CUSTOM_BADGES:
-                badge_num = badge_num or next
-                if badge_num != 0:
-                    existing = self.query(Attendee).filter_by(badge_type=badge_type, badge_num=badge_num) \
-                                                   .filter(Attendee.id != attendee.id)
-                    if existing.count():
-                        return 'That badge number already belongs to {!r}'.format(existing.first().full_name)
-            else:
+            old_badge_num = int(old_badge_num or 0) or None
+
+            if old_badge_type == attendee.badge_type and (not attendee.badge_num or old_badge_num == attendee.badge_num):
+                attendee.badge_num = old_badge_num
+                return 'Attendee is already {} with badge {}'.format(c.BADGES[old_badge_type], old_badge_num)
+
+            if c.SHIFT_CUSTOM_BADGES:
                 # fill in the gap from the old number, if applicable
+                badge_num_keep = attendee.badge_num
                 if old_badge_num:
                     self.shift_badges(old_badge_type, old_badge_num + 1, down=True)
 
-                # determine the new badge number now that the badges have shifted
-                next = self.next_badge_num(badge_type, old_badge_num)
-                badge_num = min(int(badge_num) or next, next)
-
                 # make room for the new number, if applicable
-                if badge_num:
-                    offset = 1 if badge_type == old_badge_type and old_badge_num and badge_num > old_badge_num else 0
-                    self.shift_badges(badge_type, badge_num + offset, up=True)
+                if attendee.badge_num:
+                    offset = 1 if old_badge_type == attendee.badge_type and attendee.badge_num > (old_badge_num or 0) else 0
+                    self.shift_badges(attendee.badge_type, attendee.badge_num + offset, up=True)
+                attendee.badge_num = badge_num_keep
+
+            if not attendee.badge_num and self.needs_badge_num(attendee):
+                attendee.badge_num = self.get_next_badge_num(attendee.badge_type)
 
-            attendee.badge_num = badge_num
-            attendee.badge_type = badge_type
             return 'Badge updated'
 
+        def auto_badge_num(self, badge_type):
+            """
+            Gets the next available badge number for a badge type's range.
+
+            This ignores the session so that plugins can override the logic here if need be without worrying about
+            handling dirty sessions.
```

Comment:
Oh, uh, good point.

I just mean that I split one function into two because that we we can override the business logic of "what is the next badge number" without changing the code that handles "what if we haven't flushed the session yet."